New tainting_rule type and parse its yaml file

aryx opened the pull request:
This is the first step towards taint-tracking integration into semgrep

This will help issue #585

Test plan:
/home/pad/github/semgrep/semgrep-core/_build/default/bin/main_semgrep_core.exe -lang js -dump_tainting_rules data/basic_tainting.yml
[("classic", [Tag3 (Tag10 (Tag7 (["source"; (("source", 0, 1, 0, "/tmp/tmp-29556-3308e5.py"))], ((0), (0), (0))), [(Tag26 ([(("...", 7, 1, 7, "/tmp/tmp-29556-3308e5.py"))]))])); Tag3 (Tag10 (Tag7 (["source1"; (("source1", 0, 1, 0, "/tmp/tmp-29556-610dbd.py"))], ((0), (0), (0))), [(Tag26 ([(("...", 8, 1, 8, "/tmp/tmp-29556-610dbd.py"))]))]))], [Tag3 (Tag10 (Tag7 (["sanitize"; (("sanitize", 0, 1, 0, "/tmp/tmp-29556-ad5a50.py"))], ((0), (0), (0))), [(Tag26 ([(("...", 9, 1, 9, "/tmp/tmp-29556-ad5a50.py"))]))]))], [Tag3 (Tag10 (Tag7 (["sink"; (("sink", 0, 1, 0, "/tmp/tmp-29556-104d40.py"))], ((0), (0), (0))), [(Tag26 ([(("...", 5, 1, 5, "/tmp/tmp-29556-104d40.py"))]))])); Tag3 (Tag10 (Tag7 (["eval"; (("eval", 0, 1, 0, "/tmp/tmp-29556-4b39bc.py"))], ((0), (0), (0))), [(Tag26 ([(("...", 5, 1, 5, "/tmp/tmp-29556-4b39bc.py"))]))]))], "A user input source() went into a dangerous sink()", 1, [0; 3])]

aryx committed:
New tainting_rule type and parse its yaml file

This is the first step towards taint-tracking integration into semgrep

This will help issue #585

Test plan:
/home/pad/github/semgrep/semgrep-core/_build/default/bin/main_semgrep_core.exe -lang js -dump_tainting_rules data/basic_tainting.yml
[("classic", [Tag3 (Tag10 (Tag7 (["source"; (("source", 0, 1, 0, "/tmp/tmp-29556-3308e5.py"))], ((0), (0), (0))), [(Tag26 ([(("...", 7, 1, 7, "/tmp/tmp-29556-3308e5.py"))]))])); Tag3 (Tag10 (Tag7 (["source1"; (("source1", 0, 1, 0, "/tmp/tmp-29556-610dbd.py"))], ((0), (0), (0))), [(Tag26 ([(("...", 8, 1, 8, "/tmp/tmp-29556-610dbd.py"))]))]))], [Tag3 (Tag10 (Tag7 (["sanitize"; (("sanitize", 0, 1, 0, "/tmp/tmp-29556-ad5a50.py"))], ((0), (0), (0))), [(Tag26 ([(("...", 9, 1, 9, "/tmp/tmp-29556-ad5a50.py"))]))]))], [Tag3 (Tag10 (Tag7 (["sink"; (("sink", 0, 1, 0, "/tmp/tmp-29556-104d40.py"))], ((0), (0), (0))), [(Tag26 ([(("...", 5, 1, 5, "/tmp/tmp-29556-104d40.py"))]))])); Tag3 (Tag10 (Tag7 (["eval"; (("eval", 0, 1, 0, "/tmp/tmp-29556-4b39bc.py"))], ((0), (0), (0))), [(Tag26 ([(("...", 5, 1, 5, "/tmp/tmp-29556-4b39bc.py"))]))]))], "A user input source() went into a dangerous sink()", 1, [0; 3])]

DrewDennison commented on semgrep-core/core/tainting_rule.ml:
```diff
@@ -0,0 +1,47 @@
+  id: string;
+
+  (* the list below are used to express disjunction *)
+  source: pattern list; 
```
nice!

DrewDennison reviewed (approved):
LGTM

aryx committed:
* semgrep-core/bin/main_semgrep_core.ml: extract helper functions of main
loop

aryx committed:
* semgrep-core/bin/main_semgrep_core.ml: skeleton for -tainting_rules_file

test plan:
 /home/pad/github/semgrep/semgrep-core/_build/default/bin/main_semgrep_core.exe -verbose -lang python -tainting_rules_file data/basic_tainting.yml tests/TAINTING/
Parsing data/basic_tainting.yml
Analyzing tests/TAINTING/tainting.py
{ "matches": [], "errors": [], "stats": { "okfiles": 1, "errorfiles": 0 } }

ulziibay commented on semgrep-core/data/basic_tainting.yml:
```diff
@@ -0,0 +1,13 @@
+    pattern-sources:
+      - source(...)
+      - source1(...)
+    pattern-sinks:
+      - sink(...)
+      - eval(...)
```
Just to be sure, these are logical OR clauses. In addition to this, is it possible to define two sinks combined by AND clause?

For example, when we were doing path-injection, we had to make patterns like
```
$P = os.path.join(..., request.$W.get(...), ...)
...
open($P, ...)
```
where data had to flow through `os.path.join` as well as `open` to be considered potentially malicious.

aryx replied:
hmm, right now the tainting by default consider that any tainted argument taint also the return value (unless it's a sanitize function), so here by default $P will be marked as tainted

ulziibay reviewed (commented):
the yaml API makes sense to me. Just left one additional request

aryx committed:
Plug semgrep with pfff tainting

This files issue #585

Test plan:
+ /home/pad/github/semgrep/semgrep-core/_build/default/bin/main_semgrep_core.exe -lang python -tainting_rules_file data/basic_tainting.yml tests/TAINTING/
tests/TAINTING/tainting.py:2:6: the ident 'source1' is not resolved
tests/TAINTING/tainting.py:5:9: the ident 'sanitize1' is not resolved
tests/TAINTING/tainting.py:8:2: the ident 'sink1' is not resolved
{
  "matches": [
    {
      "check_id": "classic",
      "path": "tests/TAINTING/tainting.py",
      "start": { "line": 8, "col": 3, "offset": 91 },
      "end": { "line": 8, "col": 10, "offset": 98 },
      "extra": {
        "message": "A user input source() went into a dangerous sink()",
        "metavars": {}
      }
    }
  ],
  "errors": [],
  "stats": { "okfiles": 1, "errorfiles": 0 }
}

aryx committed:
* pfff: use latest pfff

brendongo commented:
@aryx how do the sinks/sanitizers/sources interact with other rule.yaml keys like pattern-either/pattern/pattern-inside etc.

aryx commented:
It's in semgrep-core, which does not know anything about the logic-based pattern combination.
Right now semgrep transforms the complex yaml file (with logic operators) into a flat list of simple patterns for semgrep-core in a simpler yaml file (see semgrep-core/core/rule.ml). We could imagine the same for the tainted rules. @nbrahms had some ideas about that.

aryx commented:
We probably need a meeting to figure things out. I've created this tainted rule yaml file format as a spin-off, just to get quickly to the core of the problem. In the future I would prefer to have it nicely integrated with the rest of semgrep, to have something like $X: tainted metavariable, or something, but it requires some language design thinking.

aryx commented:
Right now I just imagine a separate tab in semgrep.live: we have the single pattern tab, then yaml logic-based operators config file, and I would imagine another yaml tainting config file tab. That way we can quickly give this to grayson/matt/ulzii/... to play with and get feedback on where the analysis is good and better than the logic-based patterns we wrote before and where we can improve.
